packages/contracts/contracts/CommitmentChain.sol

```diff
@@ -5,39 +5,155 @@ pragma experimental ABIEncoderV2;
+    function calculateStateUpdateLeaf(dt.StateUpdate memory _stateUpdate) public pure returns (dt.StateSubtreeNode memory) {
+        dt.StateSubtreeNode memory leaf;
+        // hash the state object to get the hash value
+        leaf.hashValue = keccak256(
             abi.encodePacked(
-                _leftSibling.hashValue,
-                _leftSibling.start,
-                _rightSibling.hashValue,
-                _rightSibling.start
-            );
-        return packed;
+                _stateUpdate.stateObject.predicateAddress,
+                _stateUpdate.stateObject.data
+            )
+        );
+        // lower bound is the start of the SU
+        leaf.lowerBound = _stateUpdate.range.start;
+        return leaf;
     }
+
     function stateSubtreeParent(
         dt.StateSubtreeNode memory _leftSibling,
         dt.StateSubtreeNode memory _rightSibling
     ) public pure returns (dt.StateSubtreeNode memory) {
+        // if the left sibling is >= the right sibling, this is an invalid proof
+        require(_leftSibling.lowerBound < _rightSibling.lowerBound, 'Interval tree siblings must be ordered to have a valid parent.');
         dt.StateSubtreeNode memory parent;
+        // calculate hash as a commitment to all parent data
+        bytes32 computedHash = keccak256(
+            abi.encodePacked(
+                _leftSibling.hashValue,
+                _leftSibling.lowerBound,
+                _rightSibling.hashValue,
+                _rightSibling.lowerBound
+            )
+        );
+        parent.hashValue = computedHash;
+        // carry the lower bound from the left sibling
+        parent.lowerBound = _leftSibling.lowerBound;
+        return parent;
+    }
+
+    function verifySubtreeInclusionAndGetRoot(dt.StateUpdate memory _stateUpdate, dt.StateTreeInclusionProof memory _proof) public pure returns (bytes32) {
+        dt.StateSubtreeNode memory computedNode = calculateStateUpdateLeaf(_stateUpdate);
+        uint128 leafPosition = _proof.stateLeafPosition;
+        // all left siblings' lower bound must be increasing from the included SU.end onward
+        uint128 previousRightLowerBound = _stateUpdate.range.end;
+        for (uint8 level = 0; level < _proof.siblings.length; level++) {
+            dt.StateSubtreeNode memory siblingNode = _proof.siblings[level];
+            // the binaryPath up the tree is the leafPosition expressed in bits
+            uint8 isComputedRightSibling = getNthBitFromRight(leafPosition, level);
+            if (isComputedRightSibling == 0) {
+                // if binaryPath[level] == 0, the proof element is a right sibling
+                computedNode = stateSubtreeParent(computedNode, siblingNode);
+                // make sure the sibling does not violate interval proof lowerBound rules
+                require(previousRightLowerBound < siblingNode.lowerBound, 'No valid branch allows potential intersections with other branches.');
```

Comment:
I think this check is only useful the first time it is executed because `previousRightLowerBound` will actually be the _end_ of the left sibling's range. After that, it's set to `siblingNode.lowerBound` which should be the _start_ of the range, making this check [redundant](https://github.com/plasma-group/pigi/pull/320/files#diff-5a51d082885f7952ed4858e1be679f3dR71), right?